### PR TITLE
binary: add sha256 checksum to downloads

### LINKED ISSUE #3448: Docs: Consider adding validation of checksums or signatures to OPA install documentation
Currently the OPA [installation documentation](https://www.openpolicyagent.org/docs/latest/#running-opa) involves downloading and execution a binary program

At the moment, there's not mention in the documentation of how to validate a digital signature or checksum of the downloaded file prior to running it. If an attacker were able to compromise the system hosting the binary this could leave users open to attack (as happened with [Codecov](https://about.codecov.io/security-update/) )

A useful addition would be either to include checksums of the binary that can be validated, or to look at signing released files using something like cosign or notary.

### PR DESCRIPTION
**Pull request by johanneslarsson**
Adds sha256 checksum to all binaries

These checksums can be used to verify file integrity.

Fixes #3448

### CODE REVIEW

File: docs/content/_index.md [hunk 1]
```diff
@@ -664,6 +664,15 @@ chmod 755 ./opa
 > You can also download and run OPA via Docker. The latest stable image tag is
 > `openpolicyagent/opa:{{< current_docker_version >}}`.
 
+#### Checksums
+Checksums for all binaries are found in the download path by appending `.sha256` to the binary filename. 
+These steps would verify the macOS binary checksum.
+```shell
+curl -L -o opa_darwin_amd64 https://openpolicyagent.org/downloads/{{< current_version >}}/opa_darwin_amd64
+curl -L -o opa_darwin_amd64.sha256 https://openpolicyagent.org/downloads/{{< current_version >}}/opa_darwin_amd64.sha256
+cat opa_darwin_amd64.sha256 | shasum -c
```
Comment (srenatus) [inline docs/content/_index.md:673]: [nit] 
```suggestion
shasum -c opa_darwin_amd64.sha256
```

File: Makefile [hunk 2]
```diff
@@ -360,10 +370,15 @@ push-latest:
...
 	aws s3 cp $(RELEASE_DIR)/opa_darwin_$(GOARCH) s3://$(S3_RELEASE_BUCKET)/edge/opa_darwin_$(GOARCH)
+	aws s3 cp $(RELEASE_DIR)/opa_darwin_$(GOARCH).sha256 s3://$(S3_RELEASE_BUCKET)/edge/opa_darwin_$(GOARCH).sha256
 	aws s3 cp $(RELEASE_DIR)/opa_darwin_arm64_static s3://$(S3_RELEASE_BUCKET)/edge/opa_darwin_arm64_static
+	aws s3 cp $(RELEASE_DIR)/opa_darwin_arm64_static.sha256 s3://$(S3_RELEASE_BUCKET)/edge/opa_darwin_arm64_static.sha256
 	aws s3 cp $(RELEASE_DIR)/opa_windows_$(GOARCH).exe s3://$(S3_RELEASE_BUCKET)/edge/opa_windows_$(GOARCH).exe
+	aws s3 cp $(RELEASE_DIR)/opa_windows_$(GOARCH).exe.sha256 s3://$(S3_RELEASE_BUCKET)/edge/opa_windows_$(GOARCH).exe.sha256
 	aws s3 cp $(RELEASE_DIR)/opa_linux_$(GOARCH) s3://$(S3_RELEASE_BUCKET)/edge/opa_linux_$(GOARCH)
+	aws s3 cp $(RELEASE_DIR)/opa_linux_$(GOARCH).sha256 s3://$(S3_RELEASE_BUCKET)/edge/opa_linux_$(GOARCH).sha256
 	aws s3 cp $(RELEASE_DIR)/opa_linux_$(GOARCH)_static s3://$(S3_RELEASE_BUCKET)/edge/opa_linux_$(GOARCH)_static
+	aws s3 cp $(RELEASE_DIR)/opa_linux_$(GOARCH)_static.sha256 s3://$(S3_RELEASE_BUCKET)/edge/opa_linux_$(GOARCH)_static.sha256
```
Comment (srenatus) [inline Makefile:382]: 💭 This is getting out of hand. But it seems you cannot provide multiple files in the first argument position. Perhaps a `aws sync` command?

```
aws s3 sync $(RELEASE_DIR)/ s3://$(S3_RELEASE_BUCKET)/edge
```
⚠️ untested ⚠️

**Review by srenatus — APPROVED** [review]
✔️ So in this PR, the GHA binaries.zip already contains the checksums and they're OK.
✔️ Via [this workflow](https://github.com/open-policy-agent/opa/blob/v0.35.0/.github/workflows/post-tag.yaml#L72-L104), _those_ shasums should end up on the github release, too.

Traditionally, anything touching the release process breaks the release process. 😅 But lets merge this and deal with that when we get there.

Could you fix DCO or just squash it all into one commit, please?

File: Makefile (hunk 2 above)
Comment (srenatus) [reply Makefile:382]: It's OK if you want to leave this as-is; I can take care of it in a follow-up. Enough moving parts for one PR.

File: docs/content/_index.md (hunk 1 above)
Comment (johanneslarsson) [reply docs/content/_index.md:673]: Yes, thats better.

File: Makefile (hunk 2 above)
Comment (johanneslarsson) [reply Makefile:382]: Sync is awesome, I can change to that. How about adding `--delete` too, it would remove any remote files that don't exist "locally"?

Comment (johanneslarsson) [reply Makefile:373]: Added `--delete`

**Comment by johanneslarsson** [comment]
> ✔️ So in this PR, the GHA binaries.zip already contains the checksums and they're OK. ✔️ Via [this workflow](https://github.com/open-policy-agent/opa/blob/v0.35.0/.github/workflows/post-tag.yaml#L72-L104), _those_ shasums should end up on the github release, too.
> 
> Traditionally, anything touching the release process breaks the release process. 😅 But lets merge this and deal with that when we get there.
> 
> Could you fix DCO or just squash it all into one commit, please?

I'm quite sure I probably missed something 😅, but these changes was the obvious ones.

**Comment by johanneslarsson** [comment]
I read in the issue that checksums where wanted in the release notes, but I'm not sure how it would work.

**Comment by srenatus** [comment]
> I read in the issue that checksums where wanted in the release notes, but I'm not sure how it would work.

I suppose we could adapt the github release script (that creates the draft release for each push of a tag) to also generate a draft message that would include the checksums. It would then be the starting point, and the manually created release notes would be paste _above_ it or something.

But also we can worry about that next. (From a security perspective, I'm not sure I understand the difference between having the shasums in assets vs having them mentioned in the release notes. I think if you have the privilege to change one of them, you likely have it to change the other, too. I might be missing something, though.)

**Comment by srenatus** [comment]
Unrelated netlify failure... I've retried the deployment.

**Comment by johanneslarsson** [comment]
> > I read in the issue that checksums where wanted in the release notes, but I'm not sure how it would work.
> 
> I suppose we could adapt the github release script (that creates the draft release for each push of a tag) to also generate a draft message that would include the checksums. It would then be the starting point, and the manually created release notes would be paste _above_ it or something.
> 
> But also we can worry about that next. (From a security perspective, I'm not sure I understand the difference between having the shasums in assets vs having them mentioned in the release notes. I think if you have the privilege to change one of them, you likely have it to change the other, too. I might be missing something, though.)

Yes, I agree.

**Comment by johanneslarsson** [comment]
> Unrelated netlify failure... I've retried the deployment.

I'll re-phrase my doc change to trigger a retry

**Review by srenatus — APPROVED** [review]
Thanks again! 

I'm afraid I might not have the proper AWS keys to fix this if there's any fallout, so let's get this merged when people with 🔑 are online 😅

**Review by tsandall — APPROVED** [review]
@johanneslarsson this looks good to me. Thanks! I'll merge now and then verify the release automation is working.

**Pull request merged**